javalib/src/main/scala/java/util/Formatter.scala

```diff
@@ -574,6 +583,133 @@ final class Formatter private (private[this] var dest: Appendable,
+   *    not lend itself well to be factored with the more straightforward code
+   *    in `Double.toHexString()`.
+   */
+  private def formatHexFloatingPoint(flags: Flags, width: Int, precision: Int,
+      arg: Double): Unit = {
+
+    if (JDouble.isNaN(arg) || JDouble.isInfinite(arg)) {
+      formatNaNOrInfinite(flags, width, arg)
+    } else {
+      // Extract the raw bits from the argument
+
+      val ebits = 11 // exponent size
+      val mbits = 52 // mantissa size
+      val mbitsMask = ((1L << mbits) - 1L)
+      val bias = (1 << (ebits - 1)) - 1
+
+      val bits = JDouble.doubleToLongBits(arg)
+      val negative = bits < 0
+      val explicitMBits = bits & mbitsMask
+      val biasedExponent = (bits >>> mbits).toInt & ((1 << ebits) - 1)
+
+      // Compute the actual precision
+
+      val actualPrecision = {
+        if (precision == 0) 1 // apparently, this is how it behaves on the JVM
+        else if (precision > 12) -1 // important for subnormals
+        else precision
+      }
+
+      // Sign string
+
+      val signStr = {
+        if (negative) "-"
+        else if (flags.positivePlus) "+"
+        else if (flags.positiveSpace) " "
+        else ""
+      }
+
+      /* Extract the implicit bit, the mantissa, and the exponent.
+       * Also apply the artificial normalization of subnormals when the
+       * actualPrecision is in the interval [1, 12].
+       */
+
+      val (implicitBitStr, mantissa, exponent) = if (biasedExponent == 0) {
+        if (explicitMBits == 0L) {
+          // Zero
+          ("0", 0L, 0)
+        } else {
+          // Subnormal
+          if (actualPrecision == -1) {
+            ("0", explicitMBits, -1022)
+          } else {
+            // Artificial normalization, required by the 'a' conversion spec
+            val leadingZeros = java.lang.Long.numberOfLeadingZeros(explicitMBits)
+            val shift = (leadingZeros + 1) - (64 - mbits)
+            val normalizedMantissa = (explicitMBits << shift) & mbitsMask
+            val normalizedExponent = -1022 - shift
+            ("1", normalizedMantissa, normalizedExponent)
+          }
+        }
+      } else {
+        // Normalized
+        ("1", explicitMBits, biasedExponent - bias)
+      }
+
+      // Apply the rounding mandated by the precision
+
+      val roundedMantissa = if (actualPrecision == -1) {
+        mantissa
+      } else {
+        val roundingUnit = 1L << (mbits - (actualPrecision * 4)) // 4 bits per hex character
+        val droppedPartMask = roundingUnit - 1
+        val halfRoundingUnit = roundingUnit >> 1
+
+        val truncated = mantissa & ~droppedPartMask
+        val droppedPart = mantissa & droppedPartMask
+
+        if (droppedPart < halfRoundingUnit)
+          truncated
+        else if (droppedPart > halfRoundingUnit)
+          truncated + roundingUnit
+        else if ((truncated & roundingUnit) == 0L) // truncated is even
+          truncated
+        else
+          truncated + roundingUnit
```

Comment:
I assume round to even here is replicating JVM behavior. Maybe drop a comment about this? (especially since other floating point conversions are specified to round half up, probably because "[...] this is the rounding mode that most of us were taught in grade school.").